src/iamclient/iamclient.cpp

```diff
@@ -5,331 +5,631 @@
+    const aos::String pemCert  = request.cert().c_str();
 
-    if (const auto it = mRemoteIMs.find(nodeID); it != mRemoteIMs.cend()) {
-        auto& remoteIM = it->second;
+    LOG_DBG() << "Process apply cert request: type = " << certType;
 
-        if (!remoteIM.mChannel) {
-            remoteIM.mChannel
-                = grpc::CreateCustomChannel(remoteIM.mConfig.mURL, mCredentials, grpc::ChannelArguments());
-        }
+    aos::iam::certhandler::CertInfo certInfo;
+    aos::Error                      err = AOS_ERROR_WRAP(mProvisionManager->ApplyCert(certType, pemCert, certInfo));
+
+    return SendApplyCertResponse(nodeId, certType, certInfo.mCertURL, certInfo.mSerial, err);
+}
+
+aos::Error IAMClient::CheckCurrentNodeStatus(const std::initializer_list<aos::NodeStatus>& allowedStatuses)
+{
+    aos::NodeInfo nodeInfo;
 
-        return CertificateService::NewStub(remoteIM.mChannel);
+    auto err = mNodeInfoProvider->GetNodeInfo(nodeInfo);
+    if (!err.IsNone()) {
+        return AOS_ERROR_WRAP(err);
     }
 
-    return nullptr;
+    const bool isAllowed = std::any_of(allowedStatuses.begin(), allowedStatuses.end(),
+        [currentStatus = nodeInfo.mStatus](const aos::NodeStatus status) { return currentStatus == status; });
+
+    return !isAllowed ? AOS_ERROR_WRAP(aos::ErrorEnum::eWrongState) : aos::ErrorEnum::eNone;
 }
 
-ProvisioningServiceStubPtr IAMClient::CreateIAMProvisioningServiceStub(const std::string& nodeID)
+aos::Error IAMClient::ExecCommand(const std::string& cmdName, const std::vector<std::string>& cmdArgs)
 {
-    std::lock_guard lock(mMutex);
+    if (!cmdArgs.empty()) {
+        std::string                    output;
+        const std::vector<std::string> args {cmdArgs.begin() + 1, cmdArgs.end()};
 
-    if (const auto it = mRemoteIMs.find(nodeID); it != mRemoteIMs.cend()) {
-        auto& remoteIM = it->second;
+        auto err = ExecProcess(cmdArgs[0], args, output);
+        if (!err.IsNone()) {
+            LOG_ERR() << cmdName.c_str() << " exec failed: output = " << output.c_str()
+                      << ", err = " << AOS_ERROR_WRAP(err);
 
-        if (!remoteIM.mChannel) {
-            remoteIM.mChannel
-                = grpc::CreateCustomChannel(remoteIM.mConfig.mURL, mCredentials, grpc::ChannelArguments());
+            return err;
         }
+    }
+
+    return aos::ErrorEnum::eNone;
+}
+
+aos::Error IAMClient::ExecProcess(const std::string& cmd, const std::vector<std::string>& args, std::string& output)
+{
+    Poco::Pipe          outPipe;
+    Poco::ProcessHandle ph = Poco::Process::launch(cmd, args, nullptr, &outPipe, &outPipe);
+
+    outPipe.close(Poco::Pipe::CLOSE_WRITE);
+
+    Poco::PipeInputStream outStream(outPipe);
+    std::string           line;
 
-        return ProvisioningService::NewStub(remoteIM.mChannel);
+    while (std::getline(outStream, line)) {
+        output += line + "\n";
     }
 
-    return nullptr;
+    int exitCode = ph.wait();
```

Comment:
should `ph.wait` be before while loop? maybe create output only when error execute command?

Reply:
According to the unit tests it doesn't matter, where to wait, but the test cases had pretty short output. Currently, I don't know how it behave with a much longer output, so prefer to keep an ordinary sequence here. 
Nevertheless, I made a small code optimization here.